merge/list_test.go

```diff
@@ -0,0 +1,282 @@
+package merge_test
+
+import (
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/typed"
+
+	"sigs.k8s.io/structured-merge-diff/merge"
+)
+
+func TestInlineListMerge(t *testing.T) {
+	parser, err := typed.NewParser(`types:
+- name: lists
+  struct:
+    fields:
+    - name: items
+      type:
+        list:
+          elementType:
+            scalar: string
+          elementRelationship: associative`)
+	if err != nil {
+		t.Fatalf("Failed to create parser: %v", err)
+	}
+
+	t.Run("disjointItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- d
+- e
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("overlappingItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- a
+- b
+- d
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("disjointOverlappingItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- b
+- a
+- d
+`)
+
+		output := typed.YAMLObject(`
+items:
+- b
+- a
+- c
+- d
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling1", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- d
+- b
+- a
+`)
+
+		output := typed.YAMLObject(`
+items:
+- d
+- b
+- a
+- c
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling2", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- e
+- d
+- b
+- a
+`)
+
+		output := typed.YAMLObject(`
+items:
+- c
+- e
+- d
+- b
+- a
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling3", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- a
+- d
+- b
+- e
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- c
+- d
+- b
+- e
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderConstraintCheck", func(t *testing.T) {
```

Comment:
@apelisse @lavalamp this test is currently failing, not sure why:
`sigs.k8s.io/structured-merge-diff/merge/list_test.go:272: Wanted err = <nil>, got .items[=true]: expected string, got true`

Reply:
Yeah I don't get that error ...

Reply:
Is the yaml parser interpreting a `y` as a boolean? If you add quotes around the `y` does it work?